File: seleniumCSharp/SeleniumNunitFinal/HuynhHoangTram/Page/LoginPage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using OpenQA.Selenium;
using HuynhHoangTram.Core;

namespace HuynhHoangTram.Page
{
    public class LoginPage
    {
        private readonly WebObject _loginTitle =
            new WebObject(
                By.XPath("//h1[normalize-space(.)='Login']"),
                "Login Title"
            );

        private readonly WebObject _usernameInput =
            new WebObject(
                By.Id("userName"),
                "Username Input"
            );

        private readonly WebObject _passwordInput =
            new WebObject(
                By.Id("password"),
                "Password Input"
            );

        private readonly WebObject _loginButton =
            new WebObject(
                By.Id("login"),
                "Login Button"
            );

        private readonly WebObject _errorMessage =
            new WebObject(
                By.Id("name"),
                "Login Error Message"
            );

        public bool IsLoginPageDisplayed()
        {
            return _loginTitle.IsElementDisplayed()
                && _usernameInput.IsElementDisplayed()
                && _passwordInput.IsElementDisplayed()
                && _loginButton.IsElementDisplayed();
        }

        public void EnterUsername(string username)
        {
            _usernameInput.EnterText(username);
        }

        public void EnterPassword(string password)
        {
            _passwordInput.EnterText(password);
        }

        public void ClickLoginButton()
        {
            _loginButton.ClickOnElement();
        }

        public void Login(string username, string password)
        {
            EnterUsername(username);

            EnterPassword(password);

            ClickLoginButton();
        }

        public bool IsLoginFailed()
        {
            return _errorMessage.IsElementDisplayed();
        }

        public string GetErrorMessage()
        {
            return _errorMessage.GetText();
        }
    }
}

File: playwright/final/HoangTramHuynh/page-object/login-page.ts
import { BasePage } from './base-page';
import { Element } from '../core/element/element';
import { BASE_URL } from '../constant/url';
import { AccountData } from '../data-object/ui/account';

export class LoginPage extends BasePage {
    private readonly emailInput: Element;
    private readonly passwordInput: Element;
    private readonly loginButton: Element;

    constructor() {
        super();
        this.emailInput = new Element("//input[@name='email']");
        this.passwordInput = new Element("//input[@name='password']");
        this.loginButton = new Element("//button[@type='submit' and @value='Login']");
    }

    async openLoginPage(): Promise<void> {
        await this.openUrl(`${BASE_URL}/login`);
    }

    async enterEmail(email: string): Promise<void> {
        await this.emailInput.enterText(email);
    }

    async enterPassword(password: string): Promise<void> {
        await this.passwordInput.enterText(password);
    }

    async clickLoginButton(): Promise<void> {
        await this.loginButton.clickOnElement();
    }

    async login(accountInfo: AccountData): Promise<void> {
        await this.enterEmail(accountInfo.email);
        await this.enterPassword(accountInfo.password);
        await this.clickLoginButton();
        await this.waitForPageLoad();
        await this.waitForTimeout(6000);
    }
}